app/gui2/src/stores/imports.ts

```diff
@@ -0,0 +1,589 @@
+  imported: ImportedNames
+}
+
+export type ImportedNames = Module | List | All
+
+/** import Module.Path (as Alias)? */
+export interface Module {
+  kind: 'Module'
+  alias?: Identifier
+}
+
+/** from Module.Path import (Ident),+ */
+export interface List {
+  kind: 'List'
+  names: Identifier[]
+}
+
+/** from Module.Path import all (hiding (Ident),*)? */
+export interface All {
+  kind: 'All'
+  except: Identifier[]
+}
+
+// ========================
+// === Required imports ===
+// ========================
+
+/** Import required for the suggestion entry. */
+export type RequiredImport = QualifiedImport | UnqualifiedImport
+
+/** import Module.Path */
+export interface QualifiedImport {
+  kind: 'Qualified'
+  module: QualifiedName
+}
+
+/** from Module.Path import SomeIdentifier */
+export interface UnqualifiedImport {
+  kind: 'Unqualified'
+  from: QualifiedName
+  import: Identifier
+}
+
+/** Get the string representation of the required import statement. */
+export function requiredImportToText(value: RequiredImport): string {
+  switch (value.kind) {
+    case 'Qualified':
+      return `import ${value.module}`
+    case 'Unqualified':
+      return `from ${value.from} import ${value.import}`
+  }
+}
+
+/** A list of required imports for specific suggestion entry */
+export function requiredImports(db: SuggestionDb, entry: SuggestionEntry): RequiredImport[] {
+  switch (entry.kind) {
+    case SuggestionKind.Module:
+      return entry.reexportedIn
+        ? [
+            {
+              kind: 'Unqualified',
+              from: entry.reexportedIn,
+              import: entry.name,
+            },
+          ]
+        : [
+            {
+              kind: 'Qualified',
+              module: entryQn(entry),
+            },
+          ]
+    case SuggestionKind.Type: {
+      const from = entry.reexportedIn ? entry.reexportedIn : entry.definedIn
+      return [
+        {
+          kind: 'Unqualified',
+          from,
+          import: entry.name,
+        },
+      ]
+    }
+    case SuggestionKind.Constructor: {
+      const selfType = selfTypeEntry(db, entry)
+      return selfType ? requiredImports(db, selfType) : []
+    }
+    case SuggestionKind.Method: {
+      const isStatic = entry.selfType == null
+      const selfType = selfTypeEntry(db, entry)
+      const isExtension = selfType && selfType.definedIn !== entry.definedIn
+      const definedIn = definedInEntry(db, entry)
+      const extensionImports = isExtension && definedIn ? requiredImports(db, definedIn) : []
+      const selfTypeImports = isStatic && selfType ? requiredImports(db, selfType) : []
+      if (isStatic) {
+        return [...extensionImports, ...selfTypeImports]
+      } else {
+        return [...extensionImports]
+      }
+    }
+    case SuggestionKind.Function:
+    case SuggestionKind.Local:
+    default:
+      return []
+  }
+}
+
+function getEntryByName(db: SuggestionDb, name: QualifiedName): SuggestionEntry | undefined {
+  if (name) {
```

Comment:
`QualifiedName` cannot be empty, can it?

Reply:
Refactoring leftover, fixed.